src/golang/cmd/server/handler/edit_integration.go

```diff
@@ -0,0 +1,318 @@
+}
+
+// `updateConfig` updates `curConfigToUpdate` *in-place* with `newConfig` with
+// the same behavior as map updates.
+// It returns 3 values:
+//  - whether there's actually an update
+//  - http status code
+//  - error if there's any
+//
+// If trying to update a 'read only' field defined by `ServerToReadOnlyFieldsMap`,
+// `updateConfig` will return a 400 and an error.
+func updateConfig(
+	curConfigToUpdate map[string]string,
+	service integration.Service,
+	newConfig map[string]string,
+) (bool, int, error) {
+	readOnlyFields := serviceToReadOnlyFields[service]
+	updated := false
+	for k, v := range newConfig {
+		if v == "" {
+			continue // no update occurs
+		}
+
+		_, isReadonlyField := readOnlyFields[k]
+		curValue, existsInCurConfig := curConfigToUpdate[k]
+		if isReadonlyField && existsInCurConfig && curValue != v {
+			// Throw if:
+			// * field is read-only, and
+			// * field both exists in cur and new, and
+			// * field values are different in cur and new
+			return false, http.StatusBadRequest, errors.Newf(
+				"Error updating read-only field %s. For %s, %v are read-only fields which cannot be edited.",
+				k,
+				service,
+				readOnlyFields,
+			)
+		}
+
+		if !existsInCurConfig || curValue != v {
+			updated = true
+			curConfigToUpdate[k] = v
+		}
+	}
+
+	return updated, http.StatusOK, nil
+}
+
+func (h *EditIntegrationHandler) Prepare(r *http.Request) (interface{}, int, error) {
+	aqContext, statusCode, err := aq_context.ParseAqContext(r.Context())
+	if err != nil {
+		return nil, statusCode, errors.Wrap(err, "Unable to edit integration.")
+	}
+
+	integrationIdStr := chi.URLParam(r, routes.IntegrationIdUrlParam)
+	integrationId, err := uuid.Parse(integrationIdStr)
+	if err != nil {
+		return nil, http.StatusBadRequest, errors.Wrap(err, "Malformed integration ID.")
+	}
+
+	hasPermission, err := h.IntegrationReader.ValidateIntegrationOwnership(
+		r.Context(),
+		integrationId,
+		aqContext.OrganizationId,
+		h.Database,
+	)
+	if err != nil {
+		return nil, http.StatusBadRequest, errors.Wrap(err, "Error validating integraiton ownership.")
+	}
+
+	if !hasPermission {
+		return nil, http.StatusForbidden, ErrNoEditPermission
+	}
+
+	name, configMap, err := request.ParseIntegrationConfigFromRequest(r)
+	if err != nil {
+		return nil, http.StatusBadRequest, errors.Wrap(err, "Unable to edit integration.")
+	}
+
+	if name == integration.DemoDbIntegrationName {
+		return nil, http.StatusBadRequest, ErrEditIntegrationWithDemoName
+	}
+
+	return &EditIntegrationArgs{
+		AqContext:     aqContext,
+		IntegrationId: integrationId,
+		Name:          name,
+		UpdatedFields: configMap,
+	}, http.StatusOK, nil
+}
+
+func (h *EditIntegrationHandler) Perform(ctx context.Context, interfaceArgs interface{}) (interface{}, int, error) {
+	args := interfaceArgs.(*EditIntegrationArgs)
+	id := args.IntegrationId
+
+	emptyResp := EditIntegrationResponse{}
+
+	integrationObject, err := h.IntegrationReader.GetIntegration(ctx, id, h.Database)
```

Comment:
I think you should wrap all of the database operations in this method in a transaction. Otherwise it's possible the name gets updated but config update fails, etc.

Reply:
@saurav-c when double checking the implementation here I think we are already putting the entire `update` in one transaction so that we only commit transaction when vault update succeeded. There's no other DB writes outside the transaction